compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -12025,4 +12043,338 @@ private STNode parseByteArrayContent(SyntaxKind kind) {
+        STToken nextToken = peek();
+        if (nextToken.kind == SyntaxKind.SLASH_LT_TOKEN) {
+            return consume();
+        } else {
+            Solution sol = recover(nextToken, ParserRuleContext.SLASH_LT_TOKEN);
+            return sol.recoveredNode;
+        }
+    }
+
+    /**
+     * Parse <code> /< </code> token.
+     *
+     * @return Parsed node
+     */
+    private STNode parseDoubleSlashDoubleAsteriskLTToken() {
+        STToken nextToken = peek();
+        if (nextToken.kind == SyntaxKind.DOUBLE_SLASH_DOUBLE_ASTERISK_LT_TOKEN) {
+            return consume();
+        } else {
+            Solution sol = recover(nextToken, ParserRuleContext.DOUBLE_SLASH_DOUBLE_ASTERISK_LT_TOKEN);
+            return sol.recoveredNode;
+        }
+    }
+
+    /**
+     * Parse xml step extend list.
+     * <p>
+     * <code>xml-step-extend-list := xml-step-extend*</code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseXMLStepExtendList() {
+        List<STNode> xmlStepExtendList = new ArrayList<>();
+        STToken nextToken = peek();
+
+        // Return an empty list
+        if (isEndOfXMLStepExtend(nextToken.kind)) {
+            return STNodeFactory.createNodeList(xmlStepExtendList);
+        }
+
+        nextToken = peek();
+        STNode xmlStepExtend;
+        while (!isEndOfXMLNamePattern(nextToken.kind)) {
+            xmlStepExtend = parseXMLStepExtend();
+            xmlStepExtendList.add(xmlStepExtend);
+            nextToken = peek();
+        }
+
+        return STNodeFactory.createNodeList(xmlStepExtendList);
+    }
+
+    private boolean isEndOfXMLStepExtend(SyntaxKind tokenKind) {
+        switch (tokenKind) {
+            case DOT_LT_TOKEN:
+            case OPEN_BRACKET_TOKEN:
+            case DOT_TOKEN:
+                return false;
+            default:
+                return true;
+        }
+    }
+
+    /**
+     * Parse xml step extend list.
+     * <p>
+     * <code>
+     * xml-step-extend :=
+     *    .< xml-name-pattern >
+     *    | [ expression ]
+     *    | . method-name ( arg-list )
+     * </code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseXMLStepExtend() {
+        STToken token = peek();
+        switch (token.kind) {
+            case DOT_LT_TOKEN:
+                return parseXMLFilterExpressionRhs();
+            case DOT_TOKEN:
+                return parseMethodCallChain();
+            // token kind is already validated in isEndOfXMLStepExtend and reach here.
+            // therefore, a recover solution is not required for the default case.
+            case OPEN_BRACKET_TOKEN:
+            default:
+                return parseOpenBracketExpressionChain();
+
+        }
+    }
+
+    /**
+     * Parse open bracket expression chain.
+     * <p>
+     * <code>open-bracket-expression-chain := `[` expression `]`</code>
+     *
+     * @return Parsed node
+     */
+    private STNode parseOpenBracketExpressionChain() {
```

Comment:
Can we use `parseMemberAccessExpr`? Because it already handles the contexts as well, and I feel both grammar are the same:
`expression xml-step-start [expression]` is equivalent to both:
- `expression [expression]` and 
- `expression xml-step-start xml-step-extend`

So we can simply return if its not `.<` and parseExpressionRhs will take care of the rest. wdyt?

Reply:
Please review the last commit. 
With above change we will have following problems like following.

 `a /* .<*> [e] .b()`
earlier `[e] and .b()` are parsed as two xml-step-extends. Now `[e] .b()` is combined and make a single `method-call-expr`

`a /* [d] [e] .<*>`
earlier `[d]`, `[e]` and `.<*>` are parsed as three xml-step-extends. Now `[d] [e] .<*>` is parsed as a `xml-filter-expr` as it has `expr .<xml-name-pattern>` grammar.

Reply:
I think thats OK. 
That separation in the grammar is used to describe the runtime behaviour. For the parser, it's a one structure.